source/particle/generator/interface.cc

```diff
@@ -0,0 +1,191 @@
+
+
+namespace aspect
+{
+  namespace Particle
+  {
+    namespace Generator
+    {
+      template <int dim>
+      std::pair<types::LevelInd,Particle<dim> >
+      Interface<dim>::generate_particle(const Point<dim> &position,
+                                        const types::particle_index id) const
+      {
+        const typename parallel::distributed::Triangulation<dim>::active_cell_iterator it =
+          (GridTools::find_active_cell_around_point<> (this->get_mapping(), this->get_triangulation(), position)).first;;
+
+        //Only try to add the point if the cell it is in, is on this processor
+        if (it->is_locally_owned())
+          {
+            const Particle<dim> particle(position, id);
+            const types::LevelInd cell(it->level(), it->index());
+            return std::make_pair(cell,particle);
+          }
+
+        AssertThrow(false,
+                    ExcMessage("A particle generator plugin tried to create a particle in"
+                               "a location that is not owned by the current processor. Please "
+                               "check that your particle generator ensures correct locations "
+                               "for the generated particles."));
+
+        // Avoid warnings about missing return
+        const Particle<dim> particle;
+        const types::LevelInd cell(0,0);
+        return std::make_pair(cell,particle);
+      }
+
+      template <int dim>
+      void
+      Interface<dim>::declare_parameters (ParameterHandler &)
+      {}
+
+      template <int dim>
+      void
+      Interface<dim>::parse_parameters (ParameterHandler &)
+      {}
+
+
+// -------------------------------- Deal with registering models and automating
+// -------------------------------- their setup and selection at run time
+
+      namespace
+      {
+        std_cxx1x::tuple
+        <void *,
+        void *,
+        internal::Plugins::PluginList<Interface<2> >,
+        internal::Plugins::PluginList<Interface<3> > > registered_plugins;
+      }
+
+
+
+      template <int dim>
+      void
+      register_particle_generator (const std::string &name,
+                                   const std::string &description,
+                                   void (*declare_parameters_function) (ParameterHandler &),
+                                   Interface<dim> *(*factory_function) ())
+      {
+        std_cxx1x::get<dim>(registered_plugins).register_plugin (name,
+                                                                 description,
+                                                                 declare_parameters_function,
+                                                                 factory_function);
+      }
+
+
+      template <int dim>
+      Interface<dim> *
+      create_particle_generator (ParameterHandler &prm)
+      {
+        std::string name;
+        prm.enter_subsection ("Postprocess");
+        {
+          prm.enter_subsection ("Tracers");
```

Comment:
So what to do? Do you foresee us moving particles from one sub-iteration (Newton step, fixed point iteration) to the next? Or are we always going to just do it once at the end of the time step?

In the first case, we are missing a *lot* of infrastructure, since in each sub-step, we'd only have to move by the increment between sub-steps. In the latter case, leaving it where it is may not be the worst place.